packages/utils/telemetry-utils/src/errorLogging.ts

```diff
@@ -3,102 +3,202 @@
-        const stackFrames = stack.split("\n");
-        stackFrames.shift(); // Remove "[ErrorName]: [ErrorMessage]"
-        if (errorName !== undefined) {
-            stackFrames.unshift(errorName); // Add "[ErrorName]"
-        }
-        return stackFrames.join("\n");
-    };
-
-    const message = (typeof error?.message === "string")
-        ? error.message as string
-        : String(error);
+/** Metadata to annotate an error object when annotating or normalizing it */
+export interface FluidErrorAnnotations {
+    /** Telemetry props to log with the error */
+    props?: ITelemetryProperties;
+    /** fluidErrorCode to mention if error isn't already an IFluidErrorBase */
+    errorCodeIfNone?: string;
+}
 
-    const safeProps: { message: string; errorType?: string; stack?: string } = {
+/**
+ * Take an unknown error object and extract certain known properties to be included in a new error object.
+ * The stack is preserved, along with any safe-to-log telemetry props.
+ * @param error - An error that was presumably caught, thrown from unknown origins
+ * @param newErrorFn - callback that will create a new error given the original error's message
+ * @returns A new error object "wrapping" the given error
+ */
+export function wrapError<T>(
+    error: any,
+    newErrorFn: (m: string) => T,
+): T {
+    const {
         message,
-    };
+        stack,
+    } = extractLogSafeErrorProperties(error);
+    const props = isILoggingError(error) ? error.getTelemetryProperties() : {};
 
-    if (isRegularObject(error)) {
-        const { errorType, stack, name } = error;
+    const newError = newErrorFn(message);
+    mixinTelemetryProps(newError, props);
 
-        if (typeof errorType === "string") {
-            safeProps.errorType = errorType;
-        }
-
-        if (typeof stack === "string") {
-            const errorName = (typeof name === "string") ? name : undefined;
-            safeProps.stack = removeMessageFromStack(stack, errorName);
-        }
+    if (stack !== undefined) {
+        Object.assign(newError, { stack });
     }
 
-    return safeProps;
+    return newError;
 }
 
-/** Type guard for ILoggingError */
-export const isILoggingError = (x: any): x is ILoggingError => typeof x?.getTelemetryProperties === "function";
+/**
+ * Normalize the given error yielding a valid Fluid Error
+ * @returns The same error object passed in if possible, normalized and with any provided annotations applied
+ * @param error - The error to normalize, ideally by patching in the properties of IFluidErrorBase
+ * @param annotations - Annotations to apply to the normalized error
+ */
+export function normalizeError(
+    error: unknown,
+    annotations: FluidErrorAnnotations = {},
+): IFluidErrorBase {
+    if (isRegularObject(error) && !Object.isFrozen(error)) {
+        // We can simply annotate the error and return it
+        annotateErrorObject(error, annotations);
```

Comment:
None of the calls to annotateErrorObject() passed error code, only props. I think that's the right model, i.e. I'd expect annotateErrorObject() to only accept props and only work with known error objects. I'd suggest to change its signature not to take FluidErrorAnnotations, but only accept ITelemetryProperties, and this path only call annotateErrorObject it it has error code ( and assert that pre-condition). For everything else - to call normalizeError() and it create a new object

Reply:
I see your point, I'll consider that in a follow-up PR.